src/common/arrstr.cpp

```diff
@@ -759,47 +756,47 @@ struct wxStringFragment
 
 wxStringFragment GetFragment(wxString& text)
 {
-    static const wxRegEx reSpaceOrPunct(wxS("^([[:space:]]|[[:punct:]])+"));
-    // Limit the length to make sure the value will fit into a wxUint64
-    static const wxRegEx reDigit(wxS("^[[:digit:]]{1,19}"));
-    static const wxRegEx reLetterOrSymbol("^[^[:space:]|[:punct:]|[:digit:]]+");
-
     if ( text.empty() )
         return wxStringFragment();
 
-    wxStringFragment fragment;
-    size_t           length = 0;
+    wxStringFragment   fragment;
+    wxString::const_iterator it;
 
-    // In attempt to minimize the number of wxRegEx.Matches() calls,
-    // try to do them from the most expected to the least expected
-    // string fragment type.
-    if ( reLetterOrSymbol.Matches(text) )
-    {
-        if ( reLetterOrSymbol.GetMatch(NULL, &length) )
-        {
-            fragment.type = wxStringFragment::LetterOrSymbol;
-            fragment.text = text.Left(length);
-        }
-    }
-    else if ( reDigit.Matches(text) )
+    for ( it = text.begin(); it != text.end(); ++it )
     {
-        if ( reDigit.GetMatch(NULL, &length) )
+        const wxUniChar        ch = *it;
+        wxStringFragment::Type chType = wxStringFragment::Empty;
+
+        if ( wxIsspace(ch) || wxIspunct(ch) )
+            chType = wxStringFragment::SpaceOrPunct;
+        else if ( wxIsdigit(ch) )
+            chType = wxStringFragment::Digit;
+        else
+            chType = wxStringFragment::LetterOrSymbol;
+
+        // check if evaluating the first character
+        if ( fragment.type == wxStringFragment::Empty )
         {
-            fragment.type = wxStringFragment::Digit;
-            fragment.text = text.Left(length);
-            fragment.text.ToULongLong(&fragment.value);
+            fragment.type = chType;
+            continue;
         }
-    }
-    else if ( reSpaceOrPunct.Matches(text) )
-    {
-        if ( reSpaceOrPunct.GetMatch(NULL, &length) )
+
+        // stop processing when the current character has a different
+        // string fragment type than the previously processed characters had
+        // or when there is a sequence of digits too long to fit in wxUint64
+        if ( fragment.type != chType ||
+             (fragment.type == wxStringFragment::Digit && it - text.begin() == 20) )
```

Comment:
Very minor, but could we please define a symbolic constant for this `20`?

Reply:
Done.

And as always: If possible, could you please squash all the commits into one if merged. Thanks.

Reply:
> Also perhaps @catalinr who reviewed the original natural sort PR may have something to say here.

Sure, I'll wait for a few days before merging this.
 
> And Vadim, as always: If possible, could you please squash all the commits into one if merged. Thanks.

Will do. BTW, I recommend using `git commit --fixup` (and/or `--squash`), this makes it perfectly clear which commits must be combined and even allows Git to do it on its own (with autosquash option which is set in my `.gitconfig`).